embedded/appendable/singleapp/single_app.go

```diff
@@ -484,7 +558,28 @@ func (aof *AppendableFile) Sync() error {
 }
 
 func (aof *AppendableFile) sync() error {
-	return aof.f.Sync()
+	err := aof.flush()
+	if err != nil {
+		return err
+	}
+
+	err = aof.f.Sync()
+	if aof.retryableSync && err != nil {
+		// Buffer space is not freed when there is an error during sync
+
+		// prevent data lost when fsync fails
+		// buffered data may be re-written in following
+		// flushing and syncing calls.
+		aof.fileOffset -= int64(aof.wbufFlushedOffset)
+		aof.wbufFlushedOffset = 0
+	}
+	if aof.retryableSync && err == nil {
+		// buffer space is freed
+		aof.wbufFlushedOffset = 0
```

Comment:
I see these two fields are being reset at a few different places, can we add that as a separate function

```
// Clear frees buffer space for reuse.
func (aof *A) Clear(){
		aof.wbufFlushedOffset = 0
		aof.wbufUnwrittenOffset = 0
}

```